sdk/storage/src/shares/share_file_client.cpp

```diff
@@ -0,0 +1,877 @@
+      DownloadFileToResult ret;
+      ret.ETag = std::move(response->ETag);
+      ret.LastModified = std::move(response->LastModified);
+      ret.HttpHeaders = std::move(response->HttpHeaders);
+      ret.Metadata = std::move(response->Metadata);
+      ret.IsServerEncrypted = response->IsServerEncrypted;
+      return Azure::Core::Response<DownloadFileToResult>(
+          std::move(ret),
+          std::make_unique<Azure::Core::Http::RawResponse>(std::move(response.GetRawResponse())));
+    };
+    auto ret = returnTypeConverter(firstChunk);
+
+    // Keep downloading the remaining in parallel
+    auto downloadChunkFunc
+        = [&](int64_t offset, int64_t length, int64_t chunkId, int64_t numChunks) {
+            DownloadFileOptions chunkOptions;
+            chunkOptions.Context = options.Context;
+            chunkOptions.Offset = offset;
+            chunkOptions.Length = length;
+            auto chunk = Download(chunkOptions);
+            bodyStreamToFile(
+                *(chunk->BodyStream),
+                fileWriter,
+                offset - firstChunkOffset,
+                chunkOptions.Length.GetValue(),
+                chunkOptions.Context);
+
+            if (chunkId == numChunks - 1)
+            {
+              ret = returnTypeConverter(chunk);
+            }
+          };
+
+    int64_t remainingOffset = firstChunkOffset + firstChunkLength;
+    int64_t remainingSize = fileRangeSize - firstChunkLength;
+    int64_t chunkSize;
+    if (options.ChunkSize.HasValue())
+    {
+      chunkSize = options.ChunkSize.GetValue();
+    }
+    else
+    {
+      int64_t c_grainSize = 4 * 1024;
+      chunkSize = remainingSize / options.Concurrency;
+      chunkSize = (std::max(chunkSize, int64_t(1)) + c_grainSize - 1) / c_grainSize * c_grainSize;
+      chunkSize = std::min(chunkSize, Storage::Details::c_DownloadDefaultChunkSize);
+    }
+
+    Storage::Details::ConcurrentTransfer(
+        remainingOffset, remainingSize, chunkSize, options.Concurrency, downloadChunkFunc);
+    ret->ContentLength = fileRangeSize;
+    return ret;
+  }
+
+  Azure::Core::Response<UploadFileFromResult> FileClient::UploadFrom(
+      const uint8_t* buffer,
+      std::size_t bufferSize,
+      const UploadFileFromOptions& options) const
+  {
+    ShareRestClient::File::CreateOptions protocolLayerOptions;
+    protocolLayerOptions.XMsContentLength = bufferSize;
+    protocolLayerOptions.FileAttributes = FileAttributesToString(options.SmbProperties.Attributes);
+    if (protocolLayerOptions.FileAttributes.empty())
+    {
+      protocolLayerOptions.FileAttributes = FileAttributesToString(FileAttributes::None);
+    }
+    protocolLayerOptions.FileCreationTime = options.SmbProperties.FileCreationTime;
+    protocolLayerOptions.FileLastWriteTime = options.SmbProperties.FileLastWriteTime;
+    if (options.FilePermission.HasValue())
+    {
+      protocolLayerOptions.FilePermission = options.FilePermission.GetValue();
```

Comment:
So this `FilePermission` is nullable in convenience layer while non-nullable in protocol layer?

Reply:
Yes, `FilePermission` and `FilePermissionKey` cannot both be null, but either must be null if the other one is not.

Reply:
I think if something is nullable in convenience layer, it should also be nullable in protocol layer.

Reply:
It is not necessary that the convenience layer aligns with the protocol layer, like the FileCreationTime we discussed.